EduVPN-multiOS/Coordinators/TunnelProviderManagerCoordinator.swift

```diff
@@ -258,16 +259,53 @@ class TunnelProviderManagerCoordinator: Coordinator {
     }
     
     func loadLog(completion: ((String) -> Void)? = nil) {
-        guard let vpn = currentManager?.connection as? NETunnelProviderSession else {
+        guard let session = currentManager?.connection as? NETunnelProviderSession else {
+            completion?("")
             return
         }
-        
-        try? vpn.sendProviderMessage(OpenVPNTunnelProvider.Message.requestLog.data) { data in
-            guard let data = data, let log = String(data: data, encoding: .utf8) else {
+
+        switch session.status {
+        case .connected, .reasserting:
+            // Ask the tunnel process for the log
+            try? session.sendProviderMessage(OpenVPNTunnelProvider.Message.requestLog.data) { data in
+                guard let data = data, let log = String(data: data, encoding: .utf8) else {
+                    completion?("")
+                    return
+                }
+                completion?(log)
+            }
+        case .disconnected:
+            // Read the log file directly
+            let appGroupURL = FileManager.default.containerURL(forSecurityApplicationGroupIdentifier: appGroup)
+            let tunnelKitDebugLogFilename = "debug.log"
+            guard let debugLogURL = appGroupURL?.appendingPathComponent(tunnelKitDebugLogFilename) else {
+                completion?("")
                 return
             }
-            
-            completion?(log)
+            completion?((try? String(contentsOf: debugLogURL)) ?? "")
+        case .disconnecting:
```

Comment:
Could you group these cases and not use a `fallthrough`?

Maybe even an `@unknown default:` and describe in 1 comment why only `connected` and `disconnected` are handled.

Reply:
I'll group it into a `default` with a comment covering *disconnecting* and *connecting* states. `@unknown default` results in a warning unless all cases are explicitly handled.